consensus/ibft/snapshot_test.go

```diff
@@ -96,32 +109,254 @@ func (ap *testerAccountPool) ValidatorSet() ValidatorSet {
+
+func TestSnapshot_setupSnapshot(t *testing.T) {
+	// Current validators
+	validators := []string{"A", "B", "C", "D"}
+	// New voted validators
+	candidateValidators := []string{"E", "F"}
+
+	pool := newTesterAccountPool()
+	pool.add(validators...)
+	validatorSet := pool.ValidatorSet()
+	genesis := pool.genesis()
+
+	pool.add(candidateValidators...)
+
+	newSnapshot := func(n uint64, set ValidatorSet, votes []*Vote) *Snapshot {
+		return &Snapshot{
+			Number: n,
+			Set:    set,
+			Votes:  votes,
+		}
+	}
+
+	type snapshotData struct {
+		LastBlock uint64
+		Snapshots []*Snapshot
+	}
+	var cases = []struct {
+		name           string
+		epoch          uint64
+		headers        []mockHeader
+		savedSnapshots []*Snapshot
+		expectedResult snapshotData
+	}{
+		{
+			name:    "should create genesis",
+			headers: []mockHeader{},
+			expectedResult: snapshotData{
+				LastBlock: 0,
+				Snapshots: []*Snapshot{
+					newSnapshot(0, validatorSet, []*Vote{}),
+				},
+			},
+		},
+		{
+			name: "should load from file and advance to latest height without any update if they are in same epoch",
+			headers: []mockHeader{
+				newMockHeader(validators, mine("A")),
+				newMockHeader(validators, mine("B")),
+			},
+			savedSnapshots: []*Snapshot{
+				newSnapshot(0, validatorSet, []*Vote{}),
+			},
+			expectedResult: snapshotData{
+				LastBlock: 2,
+				Snapshots: []*Snapshot{
+					newSnapshot(0, validatorSet, []*Vote{}),
+				},
+			},
+		},
+		{
+			name: "should generate snapshot from genesis because of no snapshot file",
+			headers: []mockHeader{
+				newMockHeader(validators, mine("A")),
+				newMockHeader(validators, mine("B")),
+			},
+			savedSnapshots: nil,
+			expectedResult: snapshotData{
+				LastBlock: 2,
+				Snapshots: []*Snapshot{
+					newSnapshot(0, validatorSet, []*Vote{}),
+				},
+			},
+		},
+		{
+			name:  "should generate snapshot from beginning of current epoch because of no snapshot file",
+			epoch: 3,
+			headers: []mockHeader{
+				newMockHeader(validators, mine("A")),
+				newMockHeader(validators, mine("B")),
+				newMockHeader(validators, mine("C")),
+				newMockHeader(validators, mine("D")),
+			},
+			savedSnapshots: nil,
+			expectedResult: snapshotData{
+				LastBlock: 4,
+				Snapshots: []*Snapshot{
+					newSnapshot(3, validatorSet, []*Vote{}),
+				},
+			},
+		},
+		{
+			name:  "should recover votes from the beginning of current epoch",
+			epoch: 3,
+			headers: []mockHeader{
+				newMockHeader(validators, mine("A")),
+				newMockHeader(validators, vote("B", "F", true)),
+				newMockHeader(validators, mine("C")),
+				newMockHeader(validators, vote("D", "E", true)),
+			},
+			savedSnapshots: nil,
+			expectedResult: snapshotData{
+				LastBlock: 4,
+				Snapshots: []*Snapshot{
+					newSnapshot(3, validatorSet, []*Vote{}),
+					newSnapshot(4, validatorSet, []*Vote{{
+						Validator: pool.get("D").Address(),
+						Address:   pool.get("E").Address(),
+						Authorize: true,
+					}}),
+				},
+			},
+		},
 	}
```

Comment:
Add this test case if you can:
```suggestion
		{
			name:  "should not count votes from the beginning of current epoch as there cannot be any proposals during the checkpoint block",
			epoch: 3,
			headers: []mockHeader{
				newMockHeader(validators, mine("A")),
				newMockHeader(validators, vote("B", "F", true)),
				newMockHeader(validators, mine("C")),
				newMockHeader(validators, vote("D", "E", true)),
				newMockHeader(validators, mine("D")),
				newMockHeader(validators, vote("C", "F", true)),
				newMockHeader(validators, vote("A", "F", true)),
			},
			savedSnapshots: nil,
			expectedResult: snapshotData{
				LastBlock: 7,
				Snapshots: []*Snapshot{
					newSnapshot(6, validatorSet, []*Vote{}),
					newSnapshot(7, validatorSet, []*Vote{{
						Validator: pool.get("A").Address(),
						Address:   pool.get("F").Address(),
						Authorize: true,
					}}),
				},
			},
		},
	}
```

Reply:
I added this case and it works without any change. Thanks!